### PR TITLE
Adds downfall environment

### PR DESCRIPTION
**Pull request by edbeeching**
Remaining todos:
- [x] add onnx model
- [x] add video
- [x] add hyperparameters for training (default with 2M steps with 4 parallel envs)
- [x] add example to hub and to tests on the godot-rl main repo https://github.com/edbeeching/godot_rl_agents/pull/176

### CODE REVIEW

**Comment by Ivan-267** [comment]
Nicely made. 

I've made a quick test (not full training), it seems to be working OK and is starting to learn. 

I'll add some comments just as notes, the env is working fine as is so no alterations are necessary.

File: examples/DownFall/scripts/spawn_box.gd [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+func get_spawn_point() -> Vector3:
+	var half_extents = shape.extents / 2.0
+	var random_point = Vector3(
+		randf_range(-half_extents.x, half_extents.x),
+		randf_range(-half_extents.y, half_extents.y),
+		randf_range(-half_extents.z, half_extents.z)
+	)
+	#print(position, random_point)
+	return random_point + position + get_parent().position # perhaps use global_transform.origin instead of position
```
Comment (Ivan-267) [inline examples/DownFall/scripts/spawn_box.gd:11]:
```suggestion
	return random_point + global_position
```

It's also possible to use just global_position as a slightly shorter version. It doesn't need to be altered as it's working fine.

File: examples/DownFall/down_fall.gd [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	# reset the player's position
+	if end:
+		player.current_level = (player.current_level + 1) % len(levels)
+	var current_level = levels[player.current_level]
+
+	var spawn_box = current_level.get_node("SpawnBox")
+	var goal = current_level.get_node("EndGate")
+
+	var spawn_point = spawn_box.get_spawn_point()
+	player.position = spawn_point
```
Comment (Ivan-267) [inline examples/DownFall/down_fall.gd:27]:
```suggestion
	player.global_position = spawn_point
```

Same as previous case.

File: examples/DownFall/scripts/animated_char.gd [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+	preload("res://materials/RedChar.tres"),
+	preload("res://materials/YellowChar.tres"),
+]
+
+func play_anim(anim):
+	$AnimationPlayer.play(anim)
+
+func _ready():
+	var mesh = $"KayKit Animated Character/Skeleton3D/PrototypePete".mesh
+	var material = materials[randi() % materials.size()]
```
Comment (Ivan-267) [inline examples/DownFall/scripts/animated_char.gd:19]:
```suggestion
	var material = materials.pick_random()
```

There's a handy method for this. I'm not sure I always remember to use it.

File: examples/DownFall/scripts/fly_cam.gd [hunk 4]
```diff
@@ -0,0 +1,30 @@
+extends CharacterBody3D
+@export var look_sensitivity: float = 0.005
+@export var max_speed : float = 20.0
+func _process(_delta):
+	if !$Camera3D.is_current():
+		return
+
+	var input_dir = Input.get_vector("left", "right", "forward", "back")
```
Comment (Ivan-267) [inline examples/DownFall/scripts/fly_cam.gd:8]: Nice, just learned about this one.

**Review by Ivan-267 — APPROVED** [review]

**Comment by Ivan-267** [comment]
I just retried with the onnx and got an inference error.
Solved by renaming the .csproj and .sln files to the name of the project.

It looks good and the agents are able to solve the levels.

**Comment by edbeeching** [comment]
@Ivan-267 ah thanks. Yes, I renamed these locally. I need to push the new files.

**Pull request merged**